### PR TITLE
Split out NearestPointBase from NearestPointLayeredAverage

### LINKED ISSUE #7048: NearestPointBase
### Description of the enhancement or error report
`NearestPointLayeredAverage` needs to be split into a base class and an inherited specialization.  This will be done by templated inheritance.

### Rationale for the enhancement or information for reproducing the error
The base class would be useful for creating many other "NearestPoint" objects including `NearestPointLayeredIntegral` or anything an application developer might think of (I have my own objects I want "NearestPoint" versions of in my code).

### Identified impact
Just moving around internals... no user facing interfaces modified.

### PR DESCRIPTION
**Pull request by friedmud**
Makes it easier for developers to create their own "NearestPoint" types of objects.

closes #7048

### CODE REVIEW

**Comment by permcody** [comment]
I know you are just moving stuff around but do you want to fix up the spacing too? All of your binary operators need spacing throughout.

**Comment by friedmud** [comment]
Sure

**Comment by friedmud** [comment]
Added another commit that cleans things up...

File: framework/include/userobject/NearestPointBase.h [hunk 1]
```diff
@@ -79,51 +79,36 @@ class NearestPointBase : public ElementIntegralVariableUserObject
    * @param p The point.
    * @return The UserObject closest to p.
    */
-  UserObjectType * nearestUserObject(const Point & p) const;
+  std::shared_ptr<UserObjectType> nearestUserObject(const Point & p) const;
```
Comment (permcody) [inline framework/include/userobject/NearestPointBase.h]: Well I guess it's time to start talking about smart pointers. I just finished reading the entire new Scott Meyer's book on C++11 and now I'm not so sure we should start sprinkling shared pointers throughout our code. He has a very good read about shared pointers vs unique pointers. I can't do the whole discussion justice here but we need to chat about it as a team so I'm going to open a ticket to start the discussion (watch for it shortly). Let's hold off on merging this until after we've had our discussion.

Comment (friedmud) [inline framework/include/userobject/NearestPointBase.h]: We do need a discussion.

I will say that this is definitely an instance of needing shared pointers.  The pointers are built and held within this object and need to continue to be held within this object while other objects might also need to store pointers to those objects.

A unique pointer is definitely not the right thing to use here.

Comment (permcody) [inline framework/include/userobject/NearestPointBase.h]: It depends. If you know who the ultimate owner is, and you know the lifetime of that object. It's fine to use a unique pointer and just pass raw pointers back to people who need access to that resource. I think shared pointers are more useful when the chain of custody is unknown or changes. (i.e. in this example does the ownership of the nearest user object really pass to the other user?)

**Comment by jwpeterson** [comment]
Sorry, I should not have merged this... didn't realize there was C++11 code there.  FYI, I've added a new label to that effect...

**Pull request merged**